srtcore/group.cpp

```diff
@@ -2078,21 +2078,36 @@ vector<CUDTSocket*> CUDTGroup::recv_WaitForReadReady(const vector<CUDTSocket*>&
     // will be surely empty. This will be checked then same way as when
     // reading from every socket resulted in error.
     vector<CUDTSocket*> readReady;
-    readReady.reserve(sready.size());
-    for (CEPoll::fmap_t::const_iterator i = sready.begin(); i != sready.end(); ++i)
+    readReady.reserve(aliveMembers.size());
+    for (vector<CUDTSocket*>::const_iterator sockiter = aliveMembers.begin(); sockiter != aliveMembers.end(); ++sockiter)
     {
-        if (i->second & SRT_EPOLL_ERR)
-            continue; // broken already
+        CUDTSocket* sock = *sockiter;
+        const CEPoll::fmap_t::const_iterator ready_iter = sready.find(sock->m_SocketID);
+        if (ready_iter != sready.end())
+        {
+            if (ready_iter->second & SRT_EPOLL_ERR)
+                continue; // broken already
+
+            if ((ready_iter->second & SRT_EPOLL_IN) == 0)
+                continue; // not ready for reading
 
-        if ((i->second & SRT_EPOLL_IN) == 0)
-            continue; // not ready for reading
+            readReady.push_back(*sockiter);
+        }
+        else
+        {
+            // No reqd-readiness on the socket, but could have missed or not yet handled, so check the state manually
+            steady_clock::time_point tsbpdtime;
+            int current_pkt_seq = 0;
 
-        // Check if this socket is in aheads
-        // If so, don't read from it, wait until the ahead is flushed.
-        SRTSOCKET   id = i->first;
-        CUDTSocket* ps = m_pGlobal->locateSocket_LOCKED(id);
-        if (ps)
-            readReady.push_back(ps);
+            int32_t skiptoseqno = SRT_SEQNO_NONE;
+            bool    passack     = true; // Get next packet to wait for even if not acked
+            const bool rxready = sock->core().m_pRcvBuffer->getRcvFirstMsg((tsbpdtime), (passack), (skiptoseqno), (current_pkt_seq));
```

Comment:
Please look into the core.cpp file, there must be a shorted function that designates the same thing.

Reply:
I don't see any. 👀

Reply:
Well, ok, but there's `CRcvBuffer::isRcvDataReady` that is shorter and easier to use.

Reply:
Good point.
It is not implemented, but my concern here is the following. What if the next packet is there, but it is not ready to be read? This can happen only if it is not yet acknowledged and thus can't be read from the buffer.
The idle link is synchronized with the latest acknowledged position of the active link - so not the case for #1757 (the next packet is acknowledged).
But it may happen in run time. Not sure what's the proper way to handle this. Force ACK?

Reply:
The whole "app-based solution" works on a premise that TSBPD is signing off the same packet coming over different links at perfectly same time. This "signing off" is an action to do, and that was the problem here, but the rules aren't changed: you have a time to play, so you play, if the time hasn't come yet, it remains in the buffer and is not allowed to be extracted. ACK means nothing here because it should have happen way before the TSBPD play time has come. Of course, if you have a situation on the edge, that is, you have a case of asymmetric loss, so both links must do TLPKTDROP, but just different packets, the reading should still happen in the order of play time.

What you are fixing here is the problem that the epoll-based read-ready state isn't synchronized with rule-based read-ready state, which is simply defined by the fact that the head-of-the-queue packet's play time is in the past, just TSBPD was not given a chance to do this synchronization. Example:
```
Link1   Link2
  |       |
  5       5
  |       |
  *       6  <---NOW
  |       |       |
  7       *       V
  |       |
  8       8
```

If we have a hypothetical situation that NOW is the moment when packet 6 is ready to play, it is being read. Next time if the reading function "wakes up" (as the call times are application-controlled), let's say its NOW time is after the play time of packet 8. You attempt to read from all ready links, but TSBPD for Link1 did not restore the read-ready state after it was cleared due to lacking packet number 6 and it's not set as well at this moment. If you call this `CRcvBuffer::isRcvDataReady` function, it will find out that at the head of the queue there's a packet number 7, which simply **cannot be not ready to play when it was already found out that packet number 8 is**, as it is guaranteed that its time to play will be less than that of packet 8.